File: Wire-iOS/Sources/Managers/SoundEventRulesWatchDog.m
// 
// Wire
// Copyright (C) 2016 Wire Swiss GmbH
// 
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with this program. If not, see http://www.gnu.org/licenses/.
// 


#import "SoundEventRulesWatchDog.h"



@implementation SoundEventRulesWatchDog

- (instancetype)initWithIgnoreTime:(NSTimeInterval)ignoreTime
{
    self = [super init];
    if (self) {
        self.ignoreTime = ignoreTime;
    }
    return self;
}

- (instancetype)init
{
    return [self initWithIgnoreTime:0.0];
}

- (BOOL)outputAllowed
{
    if (self.muted) {
        return NO;
    }
    // Otherwise check if we passed the @c ignoreTime starting from @c watchTime
    NSDate *currentTime = [NSDate date];
    NSDate *stayQuiteTillTime = [self.startIgnoreDate dateByAddingTimeInterval:self.ignoreTime];
    if ([currentTime compare:stayQuiteTillTime] == NSOrderedDescending) {
        return YES;
    }
    return NO;
}

@end

File: wire-ios/Wire-iOS/Sources/Managers/SoundEventRulesWatchDog.swift
//
// Wire
// Copyright (C) 2024 Wire Swiss GmbH
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program. If not, see http://www.gnu.org/licenses/.
//

import Foundation

final class SoundEventRulesWatchDog {
    // MARK: Lifecycle

    init(ignoreTime: TimeInterval = 0) {
        self.ignoreTime = ignoreTime
    }

    // MARK: Internal

    var startIgnoreDate: Date?
    var ignoreTime: TimeInterval
    /// Enables/disables any sound playback.
    var isMuted = false

    var outputAllowed: Bool {
        // Check this property when it is allowed to playback any sounds
        // Otherwise check if we passed the @c ignoreTime starting from @c watchTime
        guard !isMuted,
              let stayQuiteTillTime = startIgnoreDate?.addingTimeInterval(ignoreTime) else {
            return false
        }

        return Date().compare(stayQuiteTillTime) == .orderedDescending
    }
}
